### ISSUE TITLE
restricting_mutation_reader is not accurate

### ISSUE THREAD

**Issue by avikivity** [scylladb/scylla#2692]
`restricting_mutation_reader` is used to limit sstable reader concurrency, as each reader costs 128k * (number of read aheads) bytes. Enough of them and you run out of memory.

There are two problems:
 - we count the number of combined readers, but those are actually composed of multiple mutation readers (whose count can now change during the lifetime of the reader)
 - we don't keep track of the number of read-ahead segments per sstable reader (whose number also changes during the lifetime of the reader)

**Comment by avikivity** [comment]
/cc @denesb 

Implementation notes: you can track read-ahead buffers by wrapping the `data_source` that is used to feed the reader with some interposer that counts buffer allocations and deallocations.

**Comment by denesb** [comment]
So if I understand correctly the accurate method would be memory-consumption based, as that is the resource whose usage we try to restrict in fact.

We can easily track the amount of memory currently in use by existing readers, however it's more tricky to estimate how much a fresh reader would consume upon the first `operator()()` call - can we use the logic inside `file_data_source_impl::minimal_buffer_size()` for this?

Currently there are two limits:
```
    static constexpr size_t max_concurrent_reads() { return 100; }
    static constexpr size_t max_streaming_concurrent_reads() { return 10; }
```
How would this translate to memory usage? There is a comment on [max_queue_length](https://github.com/scylladb/scylla/blob/master/database.cc#L3361) saying the maximum memory that can be filled with reads is 2%, is that the magic number?

Memory usage of readers can vary to a certain degree, as `file_data_source_impl` adjust their buffer sizes and fill their read-ahead buffers so there will be an inherent inaccuracy, although still far better as the current method.

**Comment by avikivity** [comment]
`minimal_buffer_size` is the minimum read size supported by the filesystem.

You can "loan" a new reader 128kB and have it adjust afterwards based on the actual usage.  See `semaphore::consume()`.
 
2% is the magic number, it assumes an inactive reader consumes 10k. I think we can keep this assumption, and focus on the active readers, whose actual memory usage varies wildly.

If you wrap the files you give to to the reader with an interposer, you can get an accurate count of the amount of memory in use (and convert allocations to semaphore::consume(), deallocations to semaphore::signal()).

**Comment by denesb** [comment]
Do we want to keep the current separation (streaming vs. normal reads)? If yes, should I give each groups 2% of total memory or 1% (so the total is 2%)?

**Comment by avikivity** [comment]
Yes, we do.

There are two different behaviors the restriction applies. The first is to stall a reader if memory is exhausted and instead of letting it get started, queue it on the semaphore. The second is that if the memory consumed by the queue itself is too high, we reject the read and don't even queue it. This 2% is about the second behavior, and we don't want it to apply to streaming, because we're just rejecting a read that we initiated; instead the caller is supposed to limit the number of outstanding reads.

IIRC we set `restricted_mutation_reader_config::max_queue_length` to maxint to disable this second behavior for streaming.
